File: mnist_example.R
#FILE:    mnist_example.R
#AUTHOR:  Karsten Suhre
#DATE:    3 Jan 2019
#PURPOSE: evaluate usage of R with tensorflow and keras to do AI
#MODIF:   

# inspired by https://tensorflow.rstudio.com

rm(list=ls())
library(keras)

# test example from https://tensorflow.rstudio.com/keras/
mnist <- dataset_mnist()
x_train <- mnist$train$x
y_train <- mnist$train$y
x_test <- mnist$test$x
y_test <- mnist$test$y

# reshape
x_train <- array_reshape(x_train, c(nrow(x_train), 784))
x_test <- array_reshape(x_test, c(nrow(x_test), 784))

# rescale
x_train <- x_train / 255
x_test <- x_test / 255

y_train <- to_categorical(y_train, 10)
y_test <- to_categorical(y_test, 10)

# Defining the Model
model <- keras_model_sequential() 

# We begin by creating a sequential model and then adding layers using the pipe (%>%) operator
model %>% 
  layer_dense(units = 1024, activation = 'relu', input_shape = c(784)) %>% 
  layer_dropout(rate = 0.4) %>% 
  layer_dense(units = 256, activation = 'relu', input_shape = c(784)) %>% 
  layer_dropout(rate = 0.3) %>% 
  layer_dense(units = 64, activation = 'relu') %>%
  layer_dropout(rate = 0.2) %>%
  layer_dense(units = 10, activation = 'softmax')

# Next, compile the model with appropriate loss function, optimizer, and metrics:
model %>% compile(
  loss = 'categorical_crossentropy',
  optimizer = optimizer_rmsprop(),
  metrics = c('accuracy')
)

# print the model config
print(summary(model))

# Use the fit() function to train the model for N epochs using batches of batch_size images
history <- model %>% fit(
  x_train, y_train, 
  epochs = 10, batch_size = 2048, 
  #callbacks = callback_tensorboard("logs/run_a"),
  validation_split = 0.2
)

plot(history)

# Evaluate the model's performance on the test data:
eval = model %>% evaluate(x_test, y_test) 
print(eval)

File: mnist_example.py
#FILE:    mnist_example.py (python version of mnist_example.R)
#AUTHOR:  Karsten Suhre
#DATE:    6 Jan 2021
#PURPOSE: evaluate usage of python with tensorflow and keras to do AI
#MODIF:   

# run the mnist character recognition example in python
# see https://hub.packtpub.com/distributed-tensorflow-multiple-gpu-server/

import tensorflow as tf

print("Running mnist character recognition in python")
print("using tensorflow version " + tf.__version__)
print("Num GPUs Available: ", len(tf.config.list_physical_devices('GPU')))
print(tf.config.list_physical_devices('GPU'))

print("getting the data")
(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()

print("train_images.shape")
print(train_images.shape)

print("defining the model")
network = tf.keras.Sequential()

print("adding layers")
network.add(tf.keras.layers.Dense(1024, activation='relu', input_shape=(28 * 28,)))
network.add(tf.keras.layers.Dropout(rate = 0.4))
network.add(tf.keras.layers.Dense(256, activation='relu', input_shape=(28 * 28,)))
network.add(tf.keras.layers.Dropout(rate = 0.3))
network.add(tf.keras.layers.Dense(64, activation='relu', input_shape=(28 * 28,)))
network.add(tf.keras.layers.Dropout(rate = 0.2))
network.add(tf.keras.layers.Dense(10, activation='softmax'))

print("compiling the model")
network.compile(optimizer='rmsprop',
    loss='categorical_crossentropy',
    metrics=['accuracy'])

train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

train_labels = tf.keras.utils.to_categorical(train_labels)
test_labels = tf.keras.utils.to_categorical(test_labels)

print("fitting the model")
network.fit(train_images, train_labels, 
  epochs=10, batch_size=256,
  validation_split=0.2)

print("evaluating the model")
test_loss, test_acc = network.evaluate(test_images, test_labels)

print('test_acc:', test_acc)
